WIP: [cdnjs] generate static examples without api call

chris48s opened the pull request:
Having chewed over the discussion in issue #1700 and drafted a few ideas, here is one possible solution:

If we use the service class internals to produce the example from an object and split the `handle()` function into 2:

* One function which fetches data
* One function which formats it as on object that can be consumed by `_makeBadgeData()`

then it becomes easier to format an example value consistently with the real badge behviour without making any API call. If we seperate concerns correctly, this should minimise the scope for our examples to 'drift' or require a lot of maintenance.

In most cases, what we're doing fits cleanly into that pattern. In the cases where we are doing something more complicated, we can write a more complex `handle()` function, elect not to return a `staticExample` object and just return a dynamic `previewUrl` as before. If we come up with a solution that addresses the common pattern, that allows us to make an improvement over where we are now.

The one downside is that the examples obviously don't reflect reality. This will probably cause some confusion:

![gifrecord_2018-06-17_133911](https://user-images.githubusercontent.com/6025893/41508337-13f666fc-723b-11e8-8ccd-c1dcceac016b.gif)


.. but this is a problem if we implement the suggestion in issue #1700 in any way. It is not inherent to this implementation.

Thoughts? I've picked one simple example here to demonstrate, but if this approach seems sensible I can have a look at updating the examples for other badges we've already refactored, adding tests, etc..

paulmelnikow commented:
Hey @chris48s, thanks for picking this up. I really like this approach. Separating the formatting from the fetching is a smart way of separating concerns. For the current purpose it helps keep the preview badge appearances in sync. It also makes the badges more testable, more flexible, and probably easier to refactor. It enables smarter caching of service data, too.

I took a long pass through another WIP on the NPM badges, which are among the more complicated badges, and tried to get a nice feeling between the subclasses and superclasses while trying to keep a reasonably tight grip on the indirection. My work is tackling adjacent concerns, but the abstraction lines up nicely.

I wonder if we should push the abstraction up into BaseService. Almost all of our badges are doing fetch + render. That would allow the expression of these examples to be a little bit cleaner. I'll leave a few comments inline and open a PR with the work I did on NPM.

paulmelnikow commented on services/cdnjs/cdnjs.js:
```diff
@@ -7,6 +7,11 @@ const { version: versionColor} = require('../../lib/color-formatters');
 
 module.exports = class Cdnjs extends BaseJsonService {
   async handle({library}) {
```
Could you move this function declaration down? I'd prefer to declare the helper functions first, then the function that calls them. I've seen lots of work done with both styles, and find that way to be faster to read.

paulmelnikow commented on services/cdnjs/cdnjs.js:
```diff
@@ -15,8 +20,10 @@ module.exports = class Cdnjs extends BaseJsonService {
+    return json.version || 0;
+  }
 
+  static formatBadge(version) {
```
I really enjoyed working with this abstraction in #1743, where I realized how well it mapped onto React's `render()` function, which at its purest, converts the passed-in properties into a DOM-like structure. To harmonize with that, I'd suggest we pass in an object with a single named property, e.g. `static formatBadge({ version })`.

There's one thing that feels slightly off about this abstraction, which is that you want the render/formatter function to take only the properties it needs, whereas the server response might either be organized in an inconvenient way, or contain a bunch of other stuff. In the npm badges, it worked beautifully to pass a copy of the response object, pruned down to the keys used by the badge. That would work here too. For a service which returned a more poorly organized response object, we'd probably want to clean up the schema before passing it to the render/formatter function.

What do you think about using React's name, `render()`?

paulmelnikow commented on services/cdnjs/cdnjs.js:
```diff
@@ -43,7 +50,8 @@ module.exports = class Cdnjs extends BaseJsonService {
       {
-        previewUrl: 'jquery',
+        staticExample: this.formatBadge('1.5.2'),
+        exampleUrl: 'jquery',
```
If the `render` abstraction is pushed into the base class, we could put the properties here.

```js
return [{
  props: { version: '1.5.2' }, // formatBadge gets called automatically
  exampleUrl: 'jquery',
  ...
}];
```

Replacing the function call would make this easier to use with static properties when they come along.

paulmelnikow commented on services/base.js:
```diff
@@ -82,23 +82,30 @@ class BaseService {
 
+  static _makeStaticUrl(serviceData) {
+    const badgeData = this._makeBadgeData({}, serviceData);
+    const color = badgeData.colorscheme || badgeData.colorB;
```
Yay for having color support!

paulmelnikow commented on services/base.js:
```diff
@@ -82,23 +82,30 @@ class BaseService {
     return '/' + [this.url.base, partialUrl].filter(Boolean).join('/');
   }
 
+  static _makeStaticUrl(serviceData) {
```
How about `_renderToStaticBadgeUrl`?

chris48s replied:
It feels like this function and `_makeFullUrl()` are a pair of related functions and would benefit from similar names. If you're not a fan of `_makeStaticUrl()`, maybe `_makeExampleUrl()` or `_makeStaticExampleUrl()` would be better?

paulmelnikow replied:
`_makeStaticExampleUrl` works for me. There is also a function https://github.com/badges/shields/blob/66d444aa40703d401b5f13790eabd2259f136e3f/frontend/lib/badge-url.spec.js#L39-L43 which you might be able to use. (FWIW I'm not thrilled with the helpers in that module. Though I did write them, at last read they all confused me.)

chris48s commented:
Cheers for having a look at this and providing comments. I like the improved terminology `fetch()` and `render()` you've suggested.

Since you've started generalising some of these ideas into the base classes in #1743 I think it is better to park this temporarily and I'll pick this up and rebase onto master once that is in. I think that will be better than trying to work on the 2 in parallel.

chris48s commented:
Its taken a while to circle back to this one. I've updated just the one service for now as an example, but if we're happy with the tradeoffs here I'll have a look at updating the other new-style services to use static examples as well.

chris48s commented:
I've added tests and made a couple of minor tweaks.

Also, having reflected on this, I think the issue described in the first post is alleviated if we make the examples more obviously 'fake' by using placeholder values (like we already do with some badges) instead of showing a wrong badge value for a real library/repo etc. I think that is less confusing

Apart from that the main thing I've done is added static examples to existing badges. This should now cover everything we've refactored so far except for:

* NPM Type Definitions (lets review this in the context of #1897)
* Node - This one is different from all the other because the `render()` function is `async`, so I've ignored this one for now.

paulmelnikow commented on services/apm/apm.service.js:
```diff
@@ -65,6 +56,16 @@ class APMDownloads extends BaseAPMService {
+  static get examples() {
+    return [
+      {
+        exampleUrl: 'PACKAGE',
```
What do you think about using Sinatra-inspired syntax?

`:package`

This is a bit aspirational: I did a little work a while back with a library that took that kind of pattern and automatically converted it to regex, and also allowed you to substitute specific values into the pattern. That kind of thing would be really useful on the frontend.

It's also a format that's very recognizable, and attractive.

chris48s replied:
Yeah. I also thought about `{param}` as that is used by a few frameworks, but I went with the `PARAM` style as we've done that in some other places. I'm down with`:param`.

paulmelnikow commented on services/base.js:
```diff
@@ -104,17 +112,30 @@ class BaseService {
+          throw Error(
+            `Example for ${
+              this.name
+            } is missing required previewUrl or staticExample`
```
Total aside: It would be helpful if instead of printing the classname, we printed the file where the class is defined. I wonder if we can do this by introspecting in the BaseService constructor.

paulmelnikow replied:
I did something similar, to a different end, in #1934.

chris48s replied:
That's true, but I'm not sure we can use the same approach here. `caller()` returns `null` in this context

paulmelnikow replied:
Yea, that makes sense.

I was thinking there would be a way to snag it when the class was defined, though there’s no code from the superclass that runs at that time. The filename could be identified and associated with the service class at the time the service module is loaded. It’d be kind of dirty but might still be worth doing. I’ll give it some thought.

paulmelnikow commented on services/base.spec.js:
```diff
@@ -15,17 +15,26 @@ const BaseService = require('./base')
 
 class DummyService extends BaseService {
-  async handle({ namedParamA }, { queryParamA }) {
+  static render({ namedParamA, queryParamA }) {
```
Being on the reading side of this reminded me of @platan's suggestions about this test code made in https://github.com/badges/shields/pull/1589#pullrequestreview-108519067.

I think I implemented all of his suggestions except this one. Would you mind making that change while you're here?

https://github.com/badges/shields/pull/1589#discussion_r178595180

paulmelnikow commented on services/uptimerobot/uptimerobot-ratio.service.js:
```diff
@@ -24,16 +24,18 @@ module.exports = class UptimeRobotRatio extends UptimeRobotBase {
   }
 
-  static async render({ ratio }) {
+  static render({ ratio }) {
```
👍

tooomm commented:
Is there an heroku version up with the placeholders?
Or could you update the graphics in the first post?

chris48s commented:
> now that most of the examples will be placeholders: sometimes it's not completely clear how to fill out the URL, and seeing a complete working example could help a lot with that. Do you think we should keep the other examples alongside, and render them in the UI as a list of example URLs (i.e. not as live badges)?

Yeah I had a similar thought. I think it will be cluttered to show both in the 'list' view. What about if we show it in the modal dialog?

chris48s commented:
@tooomm - I've just done a staging deploy at https://shields-staging-pr-1740.herokuapp.com/ if you want to have a look. Note this is likely to change in light of above comments. The list of services where I've changed the examples is in the PR title

chris48s commented on lib/all-badge-examples.js:
```diff
@@ -340,8 +347,9 @@ const allBadgeExamples = [
         title: 'Dockbit',
         previewUri:
           '/dockbit/DockbitStatus/health.svg?token=TvavttxFHJ4qhnKstDxrvBXM',
+        placeholderUri: '/dockbit/:organisation/:pipeline.svg?token=:token',
```
Using `?token=:token` here renders this as `?token=%3Atoken` - any thoughts on a better style for this case?

paulmelnikow replied:
Heh, that's funny. My suggestion would be that we not URL-encode these. They aren't valid URLs, least I don't think they are; rather they are URL-like strings with placeholders. That justifies processing + rendering them in a different way.

chris48s commented:
I've had a go at implementing these suggestions. I've changed the terminology a bit to make things clearer, so we now have:

* PlaceholderUrl/Uri for the version with placeholders e.g: `/npm/dw/:package.svg`
* ExampleUrl/Uri for the concrete example e.g: `/npm/dw/localeval.svg`
* PreviewUrl/Uri for the static (or live) image we will actually show

For the moment in the case where we have an `ExampleUri` I've just added `ExampleUri` as a bit of text on the modal dialog. Do you have any feedback on how/where to present this? Should we actually render the badge as well as just showing it as a text example?

chris48s commented:
also preview is back up at https://shields-staging-pr-1740.herokuapp.com/

paulmelnikow commented:
> Do you have any feedback on how/where to present this? Should we actually render the badge as well as just showing it as a text example?

In the scope of this PR, I think the way you have it is great. I'd suggest one tweak, which is to place each URL on its own line, for easy copying and pasting.

Down the line we could have it so there are fields you can fill in with the values, and when you click on an example it resets the fields to match the example… 🙏 

I don't think a live badge is necessary. I'd be cool with a live link that lets you click to preview, though again, I don't think it's essential.

paulmelnikow commented on services/cdnjs/cdnjs.service.js:
```diff
@@ -59,7 +59,9 @@ module.exports = class Cdnjs extends BaseJsonService {
-        previewUrl: 'jquery',
+        placeholderUrl: ':library',
+        exampleUrl: 'jquery',
+        staticExample: this.render({ version: '1.5.2' }),
```
👍

paulmelnikow commented on services/gem/gem-rank.service.js:
```diff
@@ -76,12 +76,22 @@ module.exports = class GemRank extends BaseJsonService {
+        exampleUrl: 'rt/puppet',
+        placeholderUrl: 'rt/:package',
+        staticExample: this.render({
+          message: ordinalNumber(332),
```
It seems like the call to `ordinalNumber` is a rendering concern and belongs in `render`. Perhaps the two parameters should be `count` and a field like `interval` or `period`.

paulmelnikow commented:
This is great!

chris48s commented:
> I'd suggest one tweak, which is to place each URL on its own line, for easy copying and pasting.

Isn't it already like that..?

![screenshot at 2018-08-22 22 38 40](https://user-images.githubusercontent.com/6025893/44492397-3e772d80-a65c-11e8-99f9-a1974c6c85d8.png)

Do you mean somewhere else?

paulmelnikow commented:
That's the spot I mean.

When I triple-click the line, it selects "Example:" too:

<img width="1088" alt="screen shot 2018-08-22 at 5 47 42 pm" src="https://user-images.githubusercontent.com/1487036/44492712-86358f00-a633-11e8-8eeb-c34b4a304e6c.png">

paulmelnikow commented:
Oh, perhaps a better option would be to wrap the link in `ClickToSelect` as we're doing with the code inputs on this tab. Then a single click would select it, ready for copying.

chris48s commented:
Failures are the same issue as https://github.com/badges/shields/pull/1934#issuecomment-415143911

There's now some other badges that have been added/changed with dynamic examples that will want converting to static example. I think that's:

* appveyor tests
* npm type defs
* wercker

but the rest of the comments should be addressed now

chris48s commented on services/gem/gem-downloads.service.js:
```diff
@@ -122,22 +122,39 @@ module.exports = class GemDownloads extends BaseJsonService {
-        previewUrl: 'dv/rails/stable',
+        exampleUrl: 'dv/rails/stable',
+        urlPattern: 'dv/:package/stable',
+        staticExample: this.render({
```
I've switched to generate the labels using out existing code for doing so in 1a5d94a instead of hard-coding it.

Side note: 'stable' is not a 'tag'  in this situation. It is a special-case of version: https://github.com/badges/shields/blob/bee965b7764c1d9f15139707d1c374f2f896dc28/services/gem/gem-downloads.service.js#L78

chris48s commented:
> If you cheat and remove librariesio you could merge this.

We're only cheating ourselves ;)
I'm not going to have a chance to work on this for a few days so lets get the core stuff in now and we can continue noodling on the remaining details.

paulmelnikow commented:
This is live.
